gno.land/cmd/genesis/balances_add.go

```diff
@@ -0,0 +1,346 @@
+
+		// Skip empty lines
+		if entry == "" {
+			continue
+		}
+
+		accountBalance, err := getBalanceFromEntry(entry)
+		if err != nil {
+			return nil, fmt.Errorf("unable to extract balance data, %w", err)
+		}
+
+		balances[accountBalance.address] = accountBalance.amount
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("error encountered while scanning, %w", err)
+	}
+
+	return balances, nil
+}
+
+// getBalancesFromTransactions constructs a balance map based on MsgSend messages.
+// This way of determining the final balance sheet is not valid, since it doesn't take into
+// account different message types (ex. MsgCall) that can initialize accounts with some balance values.
+// The right way to do this sort of initialization is to spin up an in-memory node
+// and execute the entire transaction history to determine touched accounts and final balances,
+// and construct a balance sheet based off of this information
+func getBalancesFromTransactions(ctx context.Context, reader io.Reader) (accountBalances, error) {
+	balances := make(accountBalances)
+
+	scanner := bufio.NewScanner(reader)
+
+	for scanner.Scan() {
+		select {
+		case <-ctx.Done():
+			return nil, errBalanceParsingAborted
+		default:
+			// Parse the amino JSON
+			var tx std.Tx
+
+			if err := amino.UnmarshalJSON(scanner.Bytes(), &tx); err != nil {
+				return nil, fmt.Errorf(
+					"unable to unmarshal amino JSON, %w",
+					err,
+				)
+			}
+
+			feeAmount, err := getAmountFromEntry(tx.Fee.GasFee.String())
+			if err != nil {
+				return nil, fmt.Errorf(
+					"invalid gas fee amount, %s",
+					tx.Fee.GasFee.String(),
+				)
+			}
+
+			for _, msg := range tx.Msgs {
+				if msg.Type() != "send" {
+					continue
+				}
+
+				msgSend := msg.(bank.MsgSend)
+
+				sendAmount, err := getAmountFromEntry(msgSend.Amount.String())
+				if err != nil {
+					return nil, fmt.Errorf(
+						"%s, %s",
+						"invalid send amount",
+						msgSend.Amount.String(),
+					)
+				}
+
+				// This way of determining final account balances is not really valid,
+				// because we take into account only the ugnot transfer messages (MsgSend)
+				// and not other message types (like MsgCall), that can also
+				// initialize accounts with some balances. Because of this,
+				// we can run into a situation where a message send amount or fee
+				// causes an accounts balance to go < 0. In these cases,
+				// we initialize the account (it is present in the balance sheet), but
+				// with the balance of 0
+				from := balances[msgSend.FromAddress]
+				to := balances[msgSend.ToAddress]
+
+				to += sendAmount
+
+				if from < sendAmount || from < feeAmount {
+					// Account cannot cover send amount / fee
+					// (see message above)
+					from = 0
+				}
+
+				if from > sendAmount {
+					from -= sendAmount
+				}
+
+				if from > feeAmount {
+					from -= feeAmount
+				}
+
+				balances[msgSend.FromAddress] = from
+				balances[msgSend.ToAddress] = to
```

Comment:
why calculate valances? most of the time, you won't have all the info needed for that.

Reply:
The goal of this mode is to construct a "final" balance sheet from transactions.

This is obviously not 100% true, because we only parse MsgSends, and don't even take into account genesis transactions that could initialize accounts with some balance info. So this mode is essentially a _somewhat final_ balance sheet.

This mode is intended to parse blockchain tx data dumps (like we have on the tx-exports repo), and figure out some relative account balance from them